webpushd should reject messages from newer protocol versions

bnham opened the pull request:
#### e0c0585c06faa0154cfb798bd3d59288c728098f
<pre>
webpushd should reject messages from newer protocol versions
<a href="https://bugs.webkit.org/show_bug.cgi?id=242770">https://bugs.webkit.org/show_bug.cgi?id=242770</a>

Reviewed by Brady Eidson.

For forward-compatibility reasons (e.g. new STP communicating with old webpushd that was installed
with the OS), webpushd may receive messages from clients with a newer IPC protocol version. We
currently log an error and ignore such messages. We should additionally abort those connections
since we don&apos;t want to support new clients communicating with old webpushd daemons.

* Source/WebKit/webpushd/WebPushDaemon.mm:
(WebPushD::Daemon::decodeAndHandleMessage):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebPushDaemon.mm:
(TestWebKitAPI::addMessageHeaders):
(TestWebKitAPI::sendMessageToDaemon):
(TestWebKitAPI::sendMessageToDaemonWithReplySync):
(TestWebKitAPI::sendRawMessageToDaemonWithReplySync):
(TestWebKitAPI::sendConfigurationWithAuditToken):

Canonical link: <a href="https://commits.webkit.org/252492@main">https://commits.webkit.org/252492@main</a>
</pre>

cdumez commented on Source/WebKit/NetworkProcess/Notifications/Cocoa/WebPushDaemonConnectionCocoa.mm:
```diff
@@ -54,7 +54,22 @@ static void addVersionAndEncodedMessageToDictionary(Vector<uint8_t>&& message, x
+        Daemon::Decoder decoder({ static_cast<const uint8_t*>(data), dataSize });
+        std::optional<uint64_t> maybeVersion;
+        decoder >> maybeVersion;
+        RELEASE_LOG(Push, "Established connection to webpushd with protocol version: %llu", maybeVersion ? *maybeVersion : 0);
```
`maybeVersion.value_or(0)`

bnham replied:
@beidson says he prefers to not return the protocol version so I removed all of this logic from the latest PR.

cdumez commented on Source/WebKit/webpushd/WebPushDaemon.mm:
```diff
@@ -390,12 +398,14 @@ void handleWebPushDMessage(ClientConnection* connection, Span<const uint8_t> enc
-        // FIXME: Cut off this connection
+    auto version = xpc_dictionary_get_uint64(request, protocolVersionKey);
+    if (version != protocolVersionValue) {
+        RELEASE_LOG(Push, "Received request with protocol version: %llu not matching daemon protocol version: %llu", version, protocolVersionValue);
```
Probably want a RELEASE_LOG_ERROR?

cdumez reviewed (commented):
Will let someone more familiar with our push API do the formal review. Just had a few minor comments.

bnham committed:
webpushd should reject messages from newer protocol versions
https://bugs.webkit.org/show_bug.cgi?id=242770

Reviewed by Brady Eidson.

For forward-compatibility reasons (e.g. new STP communicating with old webpushd that was installed
with the OS), webpushd may receive messages from clients with a newer IPC protocol version. We
currently log an error and ignore such messages. We should additionally abort those connections
since we don't want to support new clients communicating with old webpushd daemons.

* Source/WebKit/webpushd/WebPushDaemon.mm:
(WebPushD::Daemon::decodeAndHandleMessage):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebPushDaemon.mm:
(TestWebKitAPI::addMessageHeaders):
(TestWebKitAPI::sendMessageToDaemon):
(TestWebKitAPI::sendMessageToDaemonWithReplySync):
(TestWebKitAPI::sendRawMessageToDaemonWithReplySync):
(TestWebKitAPI::sendConfigurationWithAuditToken):

Canonical link: https://commits.webkit.org/252492@main
